### PR TITLE
Relax the type of the SyntaxNodes used in the Binder from CSharp/VisualBasicSyntaxNode to just SyntaxNode.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
ping any other @dotnet/roslyn-compiler  devs?

File: src/Compilers/Core/Portable/Syntax/InternalSyntax/SyntaxDiagnosticInfoList.cs [hunk 1]
```diff
@@ -110,27 +103,26 @@ public bool MoveNext()
 
             private void PushNodeOrToken(GreenNode node)
             {
-                var token = node as SyntaxToken;
-                if (token != null)
+                if (node.IsToken)
```
Comment (jcouv) [inline src/Compilers/Core/Portable/Syntax/InternalSyntax/SyntaxDiagnosticInfoList.cs]: Could `node` be `null`?

**Comment by jcouv** [comment]
Nice. LGTM with one question above.

**Comment by CyrusNajmabadi** [comment]
retest windows_eta_open_prtest please

File: src/Compilers/Core/Portable/Syntax/InternalSyntax/SyntaxDiagnosticInfoList.cs (hunk 1 above)
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Syntax/InternalSyntax/SyntaxDiagnosticInfoList.cs]: Nope.  PushNodeOrToken is only called if node != null.  I can add an assert to that effect though!

Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Syntax/InternalSyntax/SyntaxDiagnosticInfoList.cs]: actually i'm going to eschew hte assert.  I have a objection to code that does ```Assert(x != null);  x.something```

```x.``` is already effectively asserting that it's not null.  So the assert doesn't do anything but move the crash one line higher.

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-compiler 

As part of my SyntaxRefactoring (https://github.com/dotnet/roslyn/pull/13160) i ran into something interesting that prompted this PR.

1. the change i'm working on pulls all the syntax-list types into Core so that we don't have duplicated versions of them for C# and for VB.  
2. Because the nodes are in Core they cannot subclass CSharpSyntaxNode or VBSyntaxNode.  Instead, the lists subclass SyntaxNode itself.
3. In the Bound-Node system we have BoundNodes point at syntax nodes.
4. There are Bound-Nodes that correspond to syntax-lists (for example, BoundStatementList)
5. Because of this, it was no longer possible for BoundNode to point at CSharpSyntaxNode.  Instead, it needed to be weakened to point at a SyntaxNode instead.

Doing this i discovered that the parts of the system that used BoundNodes (and their SyntaxNodes) never needed the nodes to be CSharpSyntaxNodes.  So i've relaxed this whole part of the system.

This change can be made independently, and unblocks the work to move SyntaxLists to Core.

**Comment by CyrusNajmabadi** [comment]
Note: @cston and @jcouv  signed off on this here: https://github.com/dotnet/roslyn/pull/13165

However, i'm holding off on merging this in until we get some confirmation from @VSadov and @jaredpar that it's ok to take this now.

**Comment by CyrusNajmabadi** [comment]
Note that this PR has none of the unification work happening in the other PRs. This is solely about relaxing hte type restrictions in the binder subsystem.

**Comment by CyrusNajmabadi** [comment]
Running Ty's tool seem to indicate no overall change in perf with this PR.  So i'd like to merge it in soon.  Looking for another @dotnet/roslyn-compiler  person to take a look.

**Comment by jaredpar** [comment]
Ty's tool is still very experimental.  It shoudln't be used as validation here.  We should be using the method @VSadov said.

**Comment by gafter** [comment]
I would be very surprised to see any perf difference. In fact, I would expect hardly any IL differences except the change form the use of one class type to its base class type.

**Comment by gafter** [comment]
:shipit:

**Comment by jaredpar** [comment]
@VSadov is our performance guy here.  If he thinks the perf test needs to be run before check-in then that is what we will be going with.

**Comment by CyrusNajmabadi** [comment]
I'm totally fine with running perf tests.  My main issue is that i cannot actually find any tests to run :)

I've spent a while trying to get CompilerPerfTest project running and testing.  But it doesn't seem to want to do anything (yet).  I've contacted Ty and Kevin about this, but they're also unfamiliar with this project and don't really know how to get it reporting anything.  Stepping through the project just seems to no result in it doing much unfortunately :(

**Comment by CyrusNajmabadi** [comment]
My actual overall expectation is for perf to improve with this.  My overall change removes hundreds of casts that now become unnecessary all over the compiler stack.  We do a lot of type checking just to force the nodes into a shape that we don't actually take advantage of :)

**Comment by jaredpar** [comment]
@VSadov can you help @CyrusNajmabadi getting this suite to run?

**Comment by CyrusNajmabadi** [comment]
Perf suite ran, and no perf change occurred.  Anyone have any issue with me submitting this?  It already has two signoffs.  Tnx!

**Pull request merged**